### DISCUSSION TITLE
"Tried to change the number of overrides" when running golden tests on multiple widgets with different ProviderScopes

### DISCUSSION THREAD

**Discussion by mregnauld** [rrousselGit/riverpod#2804] [Q&A] [answered]
I'm currently writing golden tests for my app, that uses Riverpod and [Golden Toolkit](https://pub.dev/packages/golden_toolkit), and I'm facing a bug that I don't understand. When running a test (see below), I get the following error:

```
The following assertion was thrown building ProviderScope(state: ProviderScopeState#61f2f):
Tried to change the number of overrides. This is not allowed – overrides cannot be removed/added,
they can only be updated.
'package:riverpod/src/framework/container.dart':
Failed assertion: line 358 pos 7: '_debugOverridesLength == overrides.length'
```

Here is the simplified test class, that I run using the following command: `flutter test --update-goldens --tags=screenshots`:

```Dart
@Tags(["screenshots"])
import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:golden_toolkit/golden_toolkit.dart';

void main()
{
  WidgetsFlutterBinding.ensureInitialized();

  final testProvider = Provider<bool?>((ref) => null);

  Future<void> generateScreenshot({
    required WidgetTester tester,
    required Widget widget,
    required String pageName,
    required Size sizeDp,
    required double density,
    CustomPump? customPump,
  }) async
  {
    await tester.pumpWidgetBuilder(widget);
    await multiScreenGolden(
      tester,
      pageName,
      customPump: customPump,
      devices: [
        Device(
          name: "final",
          size: sizeDp,
          textScale: 1,
          devicePixelRatio: density,
        ),
      ],
    );
  }

  Widget getPage1()
  {
    return ProviderScope(
      overrides: [
        testProvider.overrideWithValue(true), // if I comment this line for example, the test crashes
      ],
      child: Container(color: Colors.green),
    );
  }

  Widget getPage2()
  {
    return ProviderScope(
      overrides: [
        testProvider.overrideWithValue(false),
      ],
      child: Container(color: Colors.amber),
    );
  }

  testGoldens("Test ProviderScopes", (WidgetTester tester) async
  {
    const density = 2.0;
    const sizeBaseDp = Size(2048 / density, 2732 / density);

    await generateScreenshot(
        tester: tester,
        widget: getPage1(),
        pageName: "page1",
        sizeDp: sizeBaseDp,
        density: density);

    await generateScreenshot(
        tester: tester,
        widget: getPage2(),
        pageName: "page2",
        sizeDp: sizeBaseDp,
        density: density);
  });
  
}
```

Basically, I want to generate 2 images of 2 different widgets (page1 and page2), each widget using its own `ProviderScope`. The test passes if the content of the two `overrides` are identical, but it crashes if they are different.

What is even more surprising, is that I did something very similar on another project and it works like a charm.

Did I miss something? How can I make my test pass with two different `overrides` contents?

Thanks.

**Comment by rrousselGit** [comment]
You cannot add/remove overrides live.

**Reply by mregnauld** [reply]
Yes, I got that, but what I don't understand here is that I create two distincts `ProviderScope`, so each one has its own `overrides` (that being said, separating the two `generateScreenshot()` calls in two different `testGoldens()` solves the issue).

But still, is there a reason why I get that problem in my original example?

Thanks.

**Reply by rrousselGit** [reply, marked as answer]
You didn't create two separate ProviderScope. The second one _updated_ the first one.

If you want two different scope, give them a different Key

**Reply by mregnauld** [reply]
Awesome, it works!
Thanks!

**Reply by mregnauld** [reply]
That being said, I now have a new problem, as explained [here](https://github.com/rrousselGit/riverpod/discussions/2808)
